{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Template Submitters Script\n",
    "\n",
    "DSSU/DIQA  \n",
    "March 26, 2024"
   ],
   "id": "d5155460-6a40-4e0e-8902-a4c534ff88ef"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "knitr::opts_chunk$set(eval = FALSE)"
   ],
   "id": "00c72d81-b6b2-421c-a647-36c8568b366a"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "**Current Run Schedule:** M, W, F @ 11:00AM  \n",
    "*This script must run prior to fuzzy matching and roster compile scripts on roster days.*\n",
    "\n",
    "Sometimes adhoc submissions are added the morning of a roster, so run time should allow some opportunity for external parties to drop new files to be included in the day’s roster.\n",
    "\n",
    "*Where do the input files come from?* This script receives patient-level sequencing information submitted by laboratories in a standard template format as .csv or .xlsx files.\n",
    "\n",
    "*What happens during the script?* The patient-level information contained in the submission files is joined to COVID-19 genomic sequencing results in the WDRS ELR Entire table. Records are matched on accession number (exact match) and specimen collection date (within a margin of 14 days). Then, entries undergo cleaning, variables of interest are generated, QA checks are performed, and the results are transformed into the sequencing roster format.\n",
    "\n",
    "*What is output from this script?*\n",
    "\n",
    "-   fuzzy_match: Records that contain demographic information but cannot be matched using clinical accession number are sent to the FUZZY_MATCHING script\n",
    "\n",
    "-   write_roster_here: Valid matched records are compiled and output to the write_roster_here folder for inclusion in the sequencing roster\n",
    "\n",
    "-   For_Review/to_process: Invalid submissions that do not have a sequence status or FAILED or LOW QUALITY are sent to For_Review/to_process folder in the manual review format\n",
    "\n",
    "-   keep_na: Records that meet the criteria for the keep_na process are appended to the existing data object to be reconsidered for matching at a later date\n",
    "\n",
    "-   Completed_Submissions/Template_Submitters/last_run.csv: This is a reference list used internally by this script (i.e. not output to external sources) to document which files were processed in the prior run. Sometimes the final chunk of code fails to delete submission files that are open in the R session. This list is updated at the end of a code run and indicates which files were processed. When the script scans the Submissions folder for new files at the beginning of a code run, it uses this reference list to complete the process of removing those files if any are lingering. Using this approach, no manual action is required to move files.\n",
    "\n",
    "At the end of the script, the processed files are moved from their locations within the Submissions folder to the Completed Submissions folder. An email receipt is also generated at this time.\n",
    "\n",
    "*Note: This process is not an ideal method for receiving records. Ultimately we hope to transition all submitters to electronic laboratory reporting (ELR), which is preferable to receiving piecemeal .csv documents. However, using a template does improve consistency among submitters that have not adopted ELR, which allows us to prepare their records for rostering in a somewhat standardized manner. The incoming .csv files can be subject to formatting or submitter error and sometimes require manual edits before the script can properly ingest them.*\n",
    "\n",
    "## Developer notes\n",
    "\n",
    "In order to process files from a new submitter, a few actions need to be taken to update this script:\n",
    "\n",
    "1.  If a new sender is validated to submit patient-level sequencing information following the template format, *the laboratory name needs to be added to the lab_vars object via the* **write_lab_vars_here.R** *script*. There is a code chunk listing all confirmed template submitters (lab_names_template), and the spelling must match that of the SUBMITTING_LAB column in the submission; simply add the new name to the list! Other scripts beyond this one use that shared object to access the acceptable values for a given field, so making a change in the lab_vars object will populate downstream.\n",
    "\n",
    "2.  If the new sender very rarely submits information, it may be appropriate to drop their new files into the Submissions/adhoc subfolder. If they are regularly sending sequencing data via SFT, a designated sub-folder should be created for that submitter in the Submissions folder. *Make sure code* **Chunk 10** *of this script includes all of the Submissions sub-folders that contribute Template Submissions for ingestion into this script.* (Upon completion of the script, files are moved from their Submissions sub-folder to a corresponding Completed Submissions sub-folder – this could be created at the same time, but won’t be used until later!)\n",
    "\n",
    "3.  Then, examine the format of the GISAID_IDs included in submissions from the new laboratory to determine what modifications should be made to the validation steps in **Chunk 14**. If the GISAID_IDs follow the standard “^(hCoV-19/)?USA/WA-.\\*/\\[\\[:digit:\\]\\]{4}\\$” regex pattern, update the list of other submitters to include any new Submissions sub-folders created in the previous step. If the GISAID_IDs do not follow the expected pattern (likely due to a text prefix or unusual number of characters), a new set of code must be written to indicate the typical GISAID_ID format for that submitter for future validation – this has already been done for Aegis, Helix, and Labcorp, so there is an example to follow! In a related step, **Chunk 17** must be updated to standardize GISAID_IDs for the new submitter if they are partial or do not follow the usual regex pattern.\n",
    "\n",
    "4.  Next, updates need to be made to **Chunk 20** where columns for the final sequencing roster are created. The SEQUENCE_LAB variable is derived from the SUBMITTING_LAB variable included in the submissions. *The variable should be updated to mutate this variable properly for the new submitter.* Note that submitters sometimes use different values in the SEQUENCE_LAB column (they’re not supposed to, but it happens), but only one value per submitter should be output. This is in part addressed by the code in Chunk 20 using str_detect to identify the lab – real-life example: str_detect(toupper(SUBMITTING_LAB), “UW VIROLOGY”) ~ “UW Virology” will catch submissions that have been sent with a SUBMITTING_LAB of “UW Virology” or “UW Virology Lab”. Similarly, str_detect(toupper(SUBMITTING_LAB), “AEGIS”) ~ “Aegis” detects submissions from “Aegis” or “Aegis Sciences Corporation”. However, multiple rows may need to be written per submitter if they use acronyms or other names that do not contain shared text – example: the United States Air Force has sent submissions with the name “USAFSA” and “United States Air Force School of Aerospace” in the SUBMITTING_LAB column; two lines of code are needed for each variant of the name, but both output the same value to roster: “USAFSA”.\n",
    "\n",
    "5.  After the script has run, files need to be moved from their sub-folders within the Submissions folder to Completed Submissions. **Chunk 30** contains code to move these files. *The name of any new sub-folders in Submissions and Completed Submissions need to be included here.*\n",
    "\n",
    "These steps may not be comprehensive, and files from new submitters may contain their own idiosyncrasies - but this should offer a decent starting place if adding a newly validated laboratory to our suite of template submitters!\n",
    "\n",
    "# Script setup\n",
    "\n",
    "## Libraries"
   ],
   "id": "f044d7fc-f972-46a0-badc-865f1ee9227e"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DBI)\n",
    "library(odbc)\n",
    "library(lubridate)\n",
    "library(tidyverse)\n",
    "library(readxl)\n",
    "library(here)\n",
    "library(fs)\n",
    "library(vroom)"
   ],
   "id": "0e4114cc-d0c4-4e9b-a3f8-86e23dd055c0"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import objects\n",
    "\n",
    "## Data objects\n",
    "\n",
    "This part of the script calls several data objects that contain information required for this process. The lab_vars object is curated by Sequencing Project script runners and serves as a master reference for validated variable input. Adding new submitters, sequencing reasons, etc should be done in this object so the changes populate across all scripts that use these variables."
   ],
   "id": "75fbd6a8-2c46-4181-921b-88d2a38e1d80"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in r_creds.RDS\n",
    "r_creds <-readRDS(file.path(Sys.getenv(\"USERPROFILE\"), \"Projects/Sequencing/Data_Objects\", \"r_creds.RDS\")) \n",
    "\n",
    "# Read in keep_na\n",
    "keep_na_running <- read_csv(\"keep_na/keep_na.csv\",\n",
    "                   col_types = cols(.default = \"c\"),\n",
    "                   na = c(\"\", \"NA\", \"N/A\")) \n",
    "\n",
    "# Read in lineages\n",
    "lineages <- read_csv(\"Data_Objects/Lineages/Lineages.csv\",\n",
    "                   col_names = TRUE,\n",
    "                   col_types = cols(.default = \"c\"),\n",
    "                   na = c(\"\", \"NA\", \"N/A\")) \n",
    "\n",
    "# Read in the Washington_cumulative file (contains records submitted by CDC contracted labs)\n",
    "wa_cdc <- read_csv(\"Submissions/CDC_Cumulative/Washington_cumulative.csv\",\n",
    "                   col_types = cols(.default = \"c\"),\n",
    "                   na = c(\"\", \"NA\", \"N/A\")) \n",
    "\n",
    "# Read in lab_vars\n",
    "lab_vars <- readRDS(\"Data_Objects/lab_variables.rds\")\n",
    "\n",
    "# Read in quality_filters\n",
    "source(file.path(Sys.getenv(\"USERPROFILE\"), \"Projects/Sequencing/Roster_scripts/quality_filters.R\"))\n",
    "\n",
    "# Remove submissions that were duplicated in Completed Submissions at the end of the last run\n",
    "last_run <- read.csv(\"Completed_Submissions/Template_Submitters/Template_Submitters_File_Completed.csv\")\n",
    "\n",
    "for (i in 1:nrow(last_run)) {\n",
    "  if (file.exists(paste0(last_run[i,1]))) { \n",
    "    file.remove(paste0(last_run[i,1]))\n",
    "  }\n",
    "}"
   ],
   "id": "a129c0ce-9157-4440-8de6-16222319cd64"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Access WDRS\n",
    "\n",
    "## Connect\n",
    "\n",
    "Establish a connection to WDRS using credentials stored in a data object."
   ],
   "id": "b914de06-de34-4cbe-b468-e1a4d8270559"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# connect\n",
    "connection <- DBI::dbConnect(odbc::odbc(), \n",
    "                             Driver = r_creds$conn_list[1], \n",
    "                             Server = r_creds$conn_list[2], \n",
    "                             Database = r_creds$conn_list[3], \n",
    "                             Trusted_connection = r_creds$conn_list[4], \n",
    "                             ApplicationIntent = r_creds$conn_list[5])"
   ],
   "id": "75244b41-1ba5-4762-b637-5c9cba0a4c58"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Query WDRS\n",
    "\n",
    "After connecting to the WDRS database, create ENTIRE and FLATTENED tables in R global environment\n",
    "\n",
    "-   **ENTIRE** table contains all WDRS SARS-CoV-2 entries\n",
    "-   **FLATTENED** table contains WDRS SARS-CoV-2 entries that have been included in the sequencing roster"
   ],
   "id": "9c1c4404-e475-44e2-b797-463e67267dfc"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# WDRS ENTIRE TABLE\n",
    "    ## Pull CASE_ID, SEQUENCE_CLINICAL_ACCESSION AND COLLECTION_DATE FROM the [DD_ELR_DD_ENTIRE] table on WDRS\n",
    "    ## This table will be joined against to pull CASE_ID's using the SEQUENCE_CLINICAL_ACCESSION\n",
    "wdrs_ent <- dbGetQuery(connection, \"\n",
    "                    SELECT Distinct CASE_ID,\n",
    "                      [FILLER__ORDER__NUM] as FILLER__ORDER__NUM__WDRS,\n",
    "                      SPECIMEN__COLLECTION__DTTM as SPECIMEN__COLLECTION__DTTM__WDRS\n",
    "                    FROM [dbo].[DD_ELR_DD_ENTIRE]\n",
    "                    WHERE CODE = 'SARS' AND STATUS != '6' \n",
    "                    \")\n",
    "\n",
    "  # transform COLLECTION_DATE to character format\n",
    "wdrs_ent$SPECIMEN__COLLECTION__DTTM__WDRS <- as.character(wdrs_ent$SPECIMEN__COLLECTION__DTTM__WDRS) \n",
    "\n",
    "# WDRS FLATTENED TABLE\n",
    "    ## Pull all SEQUENCE_ACCESSION from the [DD_GCD_COVID_19_FLATTENED] table on WDRS. (these are specimens that have already been rostered)\n",
    "    ## The final roster will be filtered against this table to remove any potential duplicates\n",
    "wdrs_flat <- dbGetQuery(connection, \"\n",
    "                    SELECT DISTINCT CDC_N_COV_2019_SEQUENCE_ACCESSION_NUMBER,\n",
    "                    CDC_N_COV_2019_SEQUENCE_CLINICAL_ACCESSION_NUMBER\n",
    "                    FROM [dbo].[DD_GCD_COVID_19_FLATTENED]\n",
    "                    \")"
   ],
   "id": "3b005489-8741-4e54-8430-3a3bf5c24a84"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create matching variables\n",
    "\n",
    "SEQUENCE_ACCESSION and SEQUENCE_CLINICAL_ACCESSION values are extracted from WDRS. These unique identifiers will be used to match genomic sequencing records to the incoming submissions with patient-level information."
   ],
   "id": "825490c7-1637-40cd-9d18-275777b4e025"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a vector containing individual SEQUENCE_ACCESSIONS\n",
    "    # This variable is located in the first column of the WDRS FLATTENED table \n",
    "\n",
    "    ## Extract comma separated values if fields have multiple SEQUENCE_ACCESSIONS\n",
    "wdrs_sa_flat_split <- unlist(str_split(wdrs_flat[[1]], \",\"))\n",
    "    \n",
    "    ## Omit any NAs\n",
    "wdrs_sa_flat_clean <- wdrs_sa_flat_split[!is.na(wdrs_sa_flat_split)] %>%\n",
    "    ## Remove \"hCoV-19/\" from the beginning of SEQUENCE_ACCESSION if it has been appended (replace with \"\")\n",
    "    str_replace(\"hCoV-19/\", \"\") %>%\n",
    "    ## Trim white space resulting from str_split and remove \" \" values  \n",
    "    str_trim(\"both\")\n",
    "\n",
    "    ## Remove any values equal to \"\"\n",
    "wdrs_sa_flat_values <- wdrs_sa_flat_clean[wdrs_sa_flat_clean != \"\"]"
   ],
   "id": "e0ea50fd-4746-4a90-8d7c-e24d5cc1601c"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a vector containing individual SEQUENCE_CLINICAL_ACCESSIONS\n",
    "    # This variable is located in the second column of the WDRS FLATTENED table\n",
    "\n",
    "    ## Extract comma separated values if fields have multiple SEQUENCE_CLINICAL_ACCESSIONS\n",
    "wdrs_sca_flat_split <- unlist(str_split(wdrs_flat[[2]], \",\"))\n",
    "\n",
    "    ## Omit any NAs\n",
    "wdrs_sca_flat_clean <- wdrs_sca_flat_split[!is.na(wdrs_sca_flat_split)] %>%\n",
    "    ## Trim white space resulting from str_split and remove \" \" values  \n",
    "    str_trim(\"both\")\n",
    "\n",
    "    ## Remove any values equal to \"\"\n",
    "wdrs_sca_flat_values <- wdrs_sca_flat_clean[wdrs_sca_flat_clean != \"\"]"
   ],
   "id": "8b3a405b-38c1-4cd1-9cf4-d723ba656be6"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ingest files\n",
    "\n",
    "## Compile submissions\n",
    "\n",
    "New laboratory submissions are dropped in their corresponding sub-folders in the Submissions folder.\n",
    "\n",
    "This code chunk examines all Submissions sub-folders that have been validated to accept Template Submissions and compiles a list of paths to all files present in these folders."
   ],
   "id": "f7b355fd-30b4-4ca6-8709-1f9a5fcff11a"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify paths to folders containing Template Submissions\n",
    "submissions <- fs::dir_ls(c(\"Submissions/Adhoc\",\n",
    "                            \"Submissions/Aegis\",\n",
    "                            \"Submissions/Altius\",\n",
    "                            \"Submissions/ASU\",\n",
    "                            \"Submissions/Boise_VA\",\n",
    "                            \"Submissions/Fulgent_Genetics\",\n",
    "                            \"Submissions/Gravity Diagnostics\",\n",
    "                            \"Submissions/Helix\",\n",
    "                            \"Submissions/Infinity_Biologix\",\n",
    "                            \"Submissions/Kaiser\",\n",
    "                            \"Submissions/Labcorp\",\n",
    "                            \"Submissions/Lauring_Lab\",\n",
    "                            \"Submissions/NW_Genomics\",\n",
    "                            \"Submissions/Quest\",\n",
    "                            \"Submissions/UnitedStatesAirForceSchoolofAerospace\",\n",
    "                            \"Submissions/UW_Virology\"),\n",
    "                          recurse = TRUE, type = \"file\"\n",
    "                          )\n",
    "\n",
    "# Create a list of metadata for the new submission dataframes\n",
    "files <- submissions %>% \n",
    "  purrr::map_if(str_sub(., -4, -1) == \"xlsx\", \n",
    "         function(x) read_xlsx(x, col_types = \"text\", na = c(\"NA\", \"N/A\", \"\", \"None\", \"none\")),\n",
    "         .else =  function(x) vroom(x, col_types = cols(.default = \"c\"), na = c(\"NA\", \"N/A\", \"\", \"None\", \"none\"))\n",
    "         )\n",
    "\n",
    "# Create a backup of the submitted files in their original state\n",
    "files_original <- files"
   ],
   "id": "ab30f202-4282-4bc2-8cd2-16b8671342fb"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove empty files\n",
    "\n",
    "This code chunk examines the new submissions and identifies any empty files for removal"
   ],
   "id": "548f474b-6fb4-4e61-9140-baec68fe5428"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Confirm that a non-zero number of submissions were received\n",
    "# Confirm that the files received contain data, not just an empty template\n",
    "    ## Occasionally, a submitter may send a file with empty rows or missing observations aside from SUBMITTING_LAB\n",
    "    ## Files without any complete observations are errors by the submitter part and should be removed\n",
    "if (length(files) > 0) {\n",
    "  # For every df in files\n",
    "  for (i in 1:length(files)) {\n",
    "    # Filter to rows where the sum of fields not missing data across all columns (1:12) is greater than 1\n",
    "    files[[i]] <- filter(files[[i]], rowSums(!is.na(files[[i]][, 1:ncol(files[[i]])])) > 1)\n",
    "  }\n",
    "}\n",
    "\n",
    "# Initialize an empty vector to hold the filepaths of submissions that do not contain any data (empty)\n",
    "empty_submissions <- c()\n",
    "# For every df in files\n",
    "for (i in 1:length(files)) {\n",
    "  # If there are 0 rows of data (file is empty)\n",
    "  if (length(files) == 0) {\n",
    "    # Append the filepath to the empty_submissions\n",
    "    empty_submissions <- c(empty_submissions, names(files[i]))\n",
    "  }\n",
    "}\n",
    "\n",
    "# Remove any empty submissions from the files object\n",
    "if (length(empty_submissions) > 0) {\n",
    "  files <- files[names(files) != empty_submissions]\n",
    "}"
   ],
   "id": "1df7f655-8fad-4c78-ba16-87e06bbfe378"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Send an email and stop here if there are no files or records to process"
   ],
   "id": "60d71ef0-99cb-496f-a0e6-1ed297e1ec06"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare email details and message\n",
    "\n",
    "  ## Note that the files object had empty files removed in the previous chunk -- therefore, it will have a length equal to 0 if there were no files received OR if there were no populated files received\n",
    "if (length(files) == 0) {\n",
    "  email_from <- \"\"\n",
    "  email_to <- \"\"\n",
    "  email_subj <- \"SEQUENCING - Template Submitters\"\n",
    "  \n",
    "## If there are no new files\n",
    "ifelse(length(files_original) == 0,\n",
    "  email_body_no_files <- paste0(\"There are no new template submitter files to process on \", today()),\n",
    "  email_body_no_files <- paste0(\"\"))\n",
    "  \n",
    "## If there are only empty files\n",
    "ifelse(length(files_original) > 0,\n",
    "  email_body_only_empty <- paste0(\"All of the template submitter files received on \", today(), \" were empty (i.e. contained no records): \\n\", \n",
    "                                  print(empty_submissions), \n",
    "                                  \"\\n\\n No records will be processed by the template submitters script today.\"),\n",
    "  email_body_only_empty <- paste0(\"\"))\n",
    "\n",
    "## Compose message\n",
    "email_body <- paste0(email_body_no_files, \"\\r\", email_body_only_empty)\n",
    "\n",
    "# Send email\n",
    "sendmailR::sendmail(from = email_from,\n",
    "                    to = email_to,\n",
    "                    subject = email_subj,\n",
    "                    msg = email_body,\n",
    "                    headers= list(\"Reply-To\" = email_from),\n",
    "                    control = list(smtpServer = \"\")\n",
    "                    )\n",
    "\n",
    "stop(\"Stop script here.\", \n",
    "     \"\\n A total of \", length(files_original), \" files were received today. \\n\", \n",
    "     length(empty_submissions), \" files were empty (i.e. contained no records).\", \n",
    "     \"\\n An email receipt has been sent.\")\n",
    "}\n",
    "\n",
    "# If there are empty files submitted alongside populated files, the script can proceed. However, we want to inform stakeholders of the empty files so they can be addressed. This message will be appended to any future emails generated by this script\n",
    "ifelse(length(empty_submissions) > 0,\n",
    "  email_body_empty_files <- paste0(length(empty_submissions), \" file(s) received today was/were empty (i.e. contained no records) and may require manual follow up: \\n\", print(empty_submissions)),\n",
    "  email_body_empty_files <- paste0(\"\"))"
   ],
   "id": "b6adc4e0-eaa8-4635-a5ef-4def9ddbd099"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All files processed beyond this point will contain a non-zero number of observations.\n",
    "\n",
    "## Remove invalid files\n",
    "\n",
    "This section of code examines the new submissions and identifies any files that do not follow the template provided to laboratories for submitting records. Files that do not contain the variables specified below, with identical column names and orders, may not meet the criteria for ingestion and could encounter errors with subsequent steps in this script to prepare them for rostering. These files likely need manual review and may require corrections by the script runner or laboratory submitter.\n",
    "\n",
    "### Define valid submission criteria\n",
    "\n",
    "The following chunk establishes the parameters expected of submission files following the template. The incoming files will be compared against these criteria to determine whether they can be processed by the remainder of the script as-is."
   ],
   "id": "adc7700b-d926-4a06-8b4f-f49aac0c946a"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a vector containing the template submission variable names in the correct order\n",
    "template_col_names <- c(\n",
    "  \"LAB_ACCESSION_ID\", \n",
    "  \"GISAID_ID\", \n",
    "  \"SPECIMEN_COLLECTION_DATE\", \n",
    "  \"SUBMITTING_LAB\", \n",
    "  \"SEQUENCE_REASON\", \n",
    "  \"SEQUENCE_STATUS\", \n",
    "  \"PANGO_LINEAGE\", \n",
    "  \"FIRST_NAME\", \n",
    "  \"LAST_NAME\", \n",
    "  \"MIDDLE_NAME\", \n",
    "  \"DOB\", \n",
    "  \"ALTERNATIVE_ID\"\n",
    ")\n",
    "\n",
    "# Create a vector naming validated template submitters (SUBMITTING_LAB)\n",
    "submitting_lab_values <- lab_vars$lab_names_template\n",
    "\n",
    "# Create a vector containing valid values for SEQUENCE_REASON\n",
    "  # Note that NA and blanks are accepted\n",
    "sequence_reason_values <- lab_vars$seq_reason_template\n",
    "\n",
    "# Create a vector containing valid values for SEQUENCE_STATUS\n",
    "  # Note that NA and blanks are not accepted\n",
    "sequence_status_values <- lab_vars$seq_status_template\n",
    "\n",
    "# Extract a list of valid lineages from the lineages_list object (this should be updated daily via an automated script)\n",
    "  # Append \"\" and NA, as records with a LOW QUALITY or FAILED sequence status will not contain lineage information\n",
    "lineages_list <- c(\n",
    "  lineages[[1]], \n",
    "  \"\",\n",
    "  \"UNASSIGNED\",\n",
    "  \"Unassigned\",\n",
    "  NA\n",
    ")"
   ],
   "id": "ac788d39-e7ce-4f62-8678-a4eed3803886"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Many of the valid values are located within the lab_vars data object – this is meant to serve as a centralized source of validated information across multiple scripts, so updates to these variables only need to be made once. To add, remove, or otherwise edit relevant variables, make changes to the Roster_scripts//write_lab_variables.R script and update the data object that is output.\n",
    "\n",
    "### Prepare for validation checks\n",
    "\n",
    "Invalid files will be identified as those having formats or columns that do not match the template criteria outlined in the previous chunk.\n",
    "\n",
    "The following 8 items are checked for errors:\n",
    "\n",
    "-   *format* - variable names and order are correct\n",
    "-   *gisaid* - GISAID_ID variable is the expected format (varies by submitter) *and* there are no GISAID_IDs assigned to records with a SEQUENCE_STATUS that is not COMPLETE\n",
    "-   *lab* - SUBMITTING_LAB values are valid\n",
    "-   *reason* - SEQUENCE_REASON values are valid\n",
    "-   *status* - SEQUENCE_STATUS values are valid\n",
    "-   *collection date* - SPECIMEN_COLLECTION_DATE is in a valid format\n",
    "-   *lineage* - PANGO_LINEAGE values are valid\n",
    "\n",
    "Invalid files cannot be processed until all errors identified in this quality check are corrected.\n",
    "\n",
    "The next code chunk creates a dataframe to locate which errors are present in each submission file."
   ],
   "id": "59d33dbc-abd4-4bb7-88e6-8cdcde6374e3"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize a dataframe to identify work-stop errors within files\n",
    "invalid_submission_reason <- as.data.frame(matrix(\n",
    "  nrow = length(files),\n",
    "  ncol = 8,\n",
    "  dimnames = list(c(), c(\n",
    "    \"submission\", # identify the submission filepath\n",
    "    \"format_check\", \n",
    "    \"gisaid_check\", \n",
    "    \"lab_check\", \n",
    "    \"reason_check\", \n",
    "    \"status_check\", \n",
    "    \"coll_date_check\", \n",
    "    \"lineage_check\"))\n",
    "  )\n",
    ")\n",
    "\n",
    "# Assign each file to a row in the df\n",
    "invalid_submission_reason[,1] <- names(files)\n",
    "\n",
    "# FORMAT CHECK\n",
    "## For every submission in files, check that the formatting of the dataframe has not been altered; if so, populate error message in invalid_submissions_reason\n",
    "## Sometimes there appears to be additional columns - can use this to remove them. Replace with number of file. files[[7]] <- files[[7]][1:12]\n",
    "for (i in 1:length(files)) { \n",
    "    if(!(all(names(files[[i]]) == template_col_names))) {\n",
    "    invalid_submission_reason[i, 2] <- \"Invalid submission format - file variables do not match template\"\n",
    "  }\n",
    "}\n",
    "\n",
    "# GISAID_ID CHECK\n",
    "## GISAID_ID formats vary by submitter. for every submission in files, check that the GISAID_ID values are in the typical format for each submitter and there are not GISAID_ID values present for samples with a sequence status that is not COMPLETE\n",
    "  ## Note: Aegis, Helix, Labcorp, and Quest are currently submitting via ELR (per 03.11.2022); Because ad-hoc corrections or other records from these labs are sometimes manually dropped into Submissions, we have retained the code to process these records\n",
    "\n",
    "for (i in 1:length(files)) {\n",
    "\n",
    "  # Aegis\n",
    "  if (str_detect(names(files[i]), \"Submissions/Aegis\")) {\n",
    "    for (j in 1:nrow(files[[i]])) {\n",
    "      if (any((!(\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) == \"COMPLETE\" & str_detect(files[[i]][j,]$GISAID_ID, \"^ASC([[:digit:]])*-B.*\"))|\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) == \"LOW QUALITY\" & str_detect(files[[i]][j,]$GISAID_ID, \"^ASC([[:digit:]])*-B.*\"))|\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) != \"COMPLETE\" & is.na(files[[i]][j,]$GISAID_ID))\n",
    "      )))) {\n",
    "        invalid_submission_reason[i, 3] <- \"Invalid GISAID_IDs - check for unexpected format by submitter or GISAID_IDss assigned to records where SEQUENCE_STATUS is not COMPLETE\"\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  # Helix\n",
    "  else if (str_detect(names(files[i]), \"Submissions/Helix\")) {\n",
    "    for (j in 1:nrow(files[[i]])) {\n",
    "      if (any((!(\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) == \"COMPLETE\" & str_detect(files[[i]][j,]$GISAID_ID, \"USA/WA-CDC-STM-.*\"))|\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) == \"LOW QUALITY\" & str_detect(files[[i]][j,]$GISAID_ID, \"USA/WA-CDC-STM-.*\"))|\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) != \"COMPLETE\" & is.na(files[[i]][j,]$GISAID_ID))\n",
    "      )))) {\n",
    "        invalid_submission_reason[i, 3] <- \"Invalid GISAID_IDs - check for unexpected format by submitter or GISAID_IDss assigned to records where SEQUENCE_STATUS is not COMPLETE\"\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "\n",
    "  # Labcorp\n",
    "  else if (str_detect(names(files[i]), \"Submissions/Labcorp\")) {\n",
    "    for (j in 1:nrow(files[[i]])) {\n",
    "            if (any((!(\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) == \"COMPLETE\" & str_detect(files[[i]][j,]$GISAID_ID, \"^LC[[:digit:]]*$\"))|\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) == \"LOW QUALITY\" & str_detect(files[[i]][j,]$GISAID_ID, \"^LC[[:digit:]]*$\"))|\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) != \"COMPLETE\" & is.na(files[[i]][j,]$GISAID_ID))\n",
    "      )))) {\n",
    "        invalid_submission_reason[i, 3] <- \"Invalid GISAID_IDs - check for unexpected format by submitter or GISAID_IDss assigned to records where SEQUENCE_STATUS is not COMPLETE\"\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "\n",
    "  # Other submitters\n",
    "  else if (str_detect(names(files[i]), \"Submissions/(\n",
    "                      Adhoc|\n",
    "                      Altius|\n",
    "                      ASU|\n",
    "                      Boise_VA|\n",
    "                      Fulgent_Genetics|\n",
    "                      Gravity Diagnostics|\n",
    "                      Infinity_Biologix|\n",
    "                      Kaiser|\n",
    "                      Lauring_Lab|\n",
    "                      NW_Genomics|\n",
    "                      Quest|\n",
    "                      UnitedStatesAirForceSchoolofAerospace|\n",
    "                      UW_Virology\n",
    "                      )\")) {\n",
    "    for (j in 1:nrow(files[[i]])) {\n",
    "      if (any((!(\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) == \"COMPLETE\" & str_detect(files[[i]][j,]$GISAID_ID, \"^(hCoV-19/)?USA/WA-.*/[[:digit:]]{4}$\"))|\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) == \"LOW QUALITY\" & str_detect(files[[i]][j,]$GISAID_ID, \"^(hCoV-19/)?USA/WA-.*/[[:digit:]]{4}$\"))|\n",
    "        (toupper(files[[i]][j,]$SEQUENCE_STATUS) != \"COMPLETE\" & is.na(files[[i]][j,]$GISAID_ID))\n",
    "      )))) {\n",
    "        invalid_submission_reason[i, 3] <- \"Invalid GISAID_IDs - check for unexpected format by submitter or GISAID_IDss assigned to records where SEQUENCE_STATUS is not COMPLETE\"\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "# LAB CHECK\n",
    "## For every submission in files, check that the submitting lab variable contains valid values; if not, populate error message in invalid_submissions_reason\n",
    "    ## Note: Sometimes labs will switch the name they submit under without warning -- examples: \"UW Virology Lab\" -> \"UW Virology\", \"Aegis\" -> \"Aegis Sciences Corporation\". When a submitter makes a consistent change like this, the write_lab_vars.R script has typically been updated to accept both phrasings\n",
    "for (i in 1:length(files)) { \n",
    "  if(!(all(files[[i]]$SUBMITTING_LAB %in% submitting_lab_values))) {\n",
    "    invalid_submission_reason[i, 4] <- \"Invalid values in SUBMITTING_LAB variable\"\n",
    "  }\n",
    "}\n",
    "\n",
    "# REASON CHECK\n",
    "## For every submission in files, check that the sequence reason variable contains valid values; if not, populate error message in invalid_submissions_reason\n",
    "for (i in 1:length(files)) { \n",
    "  if(!(all(toupper(files[[i]]$SEQUENCE_REASON) %in% sequence_reason_values))) {\n",
    "    invalid_submission_reason[i, 5] <- \"Invalid values in SEQUENCE_REASON variable\"\n",
    "  }\n",
    "}\n",
    "\n",
    "# STATUS CHECK\n",
    "## For every submission in files, check that the sequence status variable contains valid values; if not, populate error message in invalid_submissions_reason\n",
    "for (i in 1:length(files)) { \n",
    "  if(!(all(toupper(files[[i]]$SEQUENCE_STATUS) %in% sequence_status_values))) {\n",
    "    invalid_submission_reason[i, 6] <- \"Invalid values in SEQUENCING_STATUS variable\"\n",
    "  }\n",
    "}\n",
    "\n",
    "# COLLECTION DATE CHECK\n",
    "## For every submission in files, check that the specimen collection date is in a valid format; if not, populate error message in invalid_submissions_reason\n",
    "for (i in 1:length(files)) { \n",
    "  if(!(all(!(is.na(files[[i]]$SPECIMEN_COLLECTION_DATE)) & \n",
    "           str_detect(files[[i]]$SPECIMEN_COLLECTION_DATE,\n",
    "                      \"[[:digit:]]{1,2}[[:punct:]][[:digit:]]{1,2}[[:punct:]][[:digit:]]{4}|[[:digit:]]{4}[[:punct:]][[:digit:]]{1,2}[[:punct:]][[:digit:]]{1,2}|[[:digit:]]{5}\")))) {\n",
    "    invalid_submission_reason[i, 7] <- \"Invalid SPECIMEN_COLLECTION_DATE - check for incorrect date format\"\n",
    "  }\n",
    "}\n",
    "\n",
    "# LINEAGE CHECK\n",
    "## For every submission in files, check that the PANGO lineage variable contains valid values; if not, populate error message in invalid_submissions_reason\n",
    "for (i in 1:length(files)) { \n",
    "  if(!(all(toupper(files[[i]]$PANGO_LINEAGE) %in% lineages_list))) {\n",
    "    invalid_submission_reason[i, 8] <- \"Invalid values in PANGO_LINEAGE variable\"\n",
    "  }\n",
    "}"
   ],
   "id": "87b3934d-c054-4e59-b84d-b6cfd292275a"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Separate valid and invalid submissions"
   ],
   "id": "9adf137a-49dd-4863-98e0-95cf8cb4ee87"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract invalid files (i.e. triggered QA check errors, do not meet the criteria to be processed)\n",
    "invalid_files_subset <- invalid_submission_reason %>%\n",
    "  filter(!(rowSums(is.na(invalid_submission_reason[, c(2:7)])) == 6))\n",
    "\n",
    "# Extract valid files (i.e. no QA check errors, meet all criteria to be processed)\n",
    "valid_files_subset <- invalid_submission_reason %>%\n",
    "  filter(rowSums(is.na(invalid_submission_reason[, c(2:7)])) == 6)\n",
    "\n",
    "# Bind valid_files into a single dataframe\n",
    "valid_files <- files[names(files) %in% valid_files_subset[[1]]]\n",
    "files_bind <- bind_rows(valid_files, .id = \"column_label\")\n",
    "\n",
    "# Built-in stop in case no valid files; Print summary\n",
    "  ## Note: sometimes a batch of files from a single submitter containing identical errors will be received, and often it just requires a simple fix. It's preferable to stop here and address the error manually rather than send out a premature email stating that all files cannot be processed today\n",
    "  ## If errors are corrected at this point, rerun from Chunk 13 (validation checks) to see if fixed! \n",
    "stop(paste0(\"NOTES FOR SCRIPT RUNNER:\",\n",
    "            \"\\n\", \"Stop script to review files.\",\n",
    "            \"\\n\\n\", length(files), \" files were detected in the Submissions folders\"),\n",
    "            \"\\n\", \"   Valid files: \", nrow(valid_files_subset),\n",
    "            \"\\n\", \"   Invalid files: \", nrow(invalid_files_subset),\n",
    "            \"\\n\\n\", \"Proceed with script or perform manual review as necessary.\")"
   ],
   "id": "1130d713-9e22-4954-999f-dc6392996961"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Stop script and send email if all incoming files are invalid."
   ],
   "id": "b20ff63e-9042-4725-84e9-bfb79eb72367"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize vectors to hold messages\n",
    "invalid_files_message <- c()\n",
    "final_invalid_files_message <- c()\n",
    "\n",
    "# For all invalid files, note the filepath and the errors that were identified\n",
    "for (i in 1:nrow(invalid_files_subset)) {\n",
    "  for (i in 1:nrow(invalid_files_subset)) {\n",
    "    invalid_files_message[i] <- paste0(invalid_files_subset[i,][!is.na(invalid_files_subset[i,])], collapse=\"\\n     - \")\n",
    "  }\n",
    "  \n",
    "  # Collapse invalid_files that have been processed by line\n",
    "  invalid_files_message <-  paste0(invalid_files_message, collapse=\"\\n\")\n",
    "  \n",
    "  # Compose final message\n",
    "  final_invalid_files_message <- paste0(\"The following file(s) are invalid and have not been processed: \",\n",
    "                                        \"\\n\", invalid_files_message)\n",
    "}\n",
    "\n",
    "# If there are files to process but none are valid (i.e all files are invalid), send an email stating so\n",
    "if ((length(files) > 0) & (length(valid_files) == 0)) {\n",
    "  email_from <- \"\"\n",
    "  email_to <- \"\"\n",
    "  email_subj <- \"SEQUENCING - Template Submitters\"\n",
    "  email_body <- paste0(\"There are submissions to process, however, the submission files are invalid and require correction. There are no valid submissions that can be processed \", today(), \".\",\n",
    "  \"\\n\", final_invalid_files_message,\n",
    "  \"\\n\\n\", email_body_empty_files) # Append message regarding any empty files\n",
    "\n",
    "# Send email\n",
    "sendmailR::sendmail(from = email_from,\n",
    "                    to = email_to,\n",
    "                    subject = email_subj,\n",
    "                    msg = email_body,\n",
    "                    headers= list(\"Reply-To\" = email_from),\n",
    "                    control = list(smtpServer = \"\")\n",
    "                    )\n",
    "}\n",
    "\n",
    "# Stop the script if all incoming submissions are invalid \n",
    "if ((length(files) > 0) & (length(valid_files) == 0)) {\n",
    "  stop(\"All files are invalid. File(s) require correction. Script will stop here.\")\n",
    "}"
   ],
   "id": "1d4d6bd3-e3a9-4ce2-a813-19ce200c8284"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Match records\n",
    "\n",
    "## Prepare for matching\n",
    "\n",
    "The next code chunk performs data cleaning and mutation to prepare the valid patient-level records for matching to WDRS entries."
   ],
   "id": "42bc2c39-3fcd-43eb-b900-c6d2870bc94e"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "records_mut <- files_bind %>%\n",
    "  \n",
    "# Standardize format of SPECIMEN_COLLECTION_DATE as ymd\n",
    "  mutate(SPECIMEN_COLLECTION_DATE = case_when(\n",
    "    # Parse as ymd when SPECIMEN_COLLECTION_DATE is a 5 digit numeric string (Excel date)\n",
    "    str_detect(SPECIMEN_COLLECTION_DATE, \"^[[:digit:]]{5}$\") ~ as.character(openxlsx::convertToDate(SPECIMEN_COLLECTION_DATE)),\n",
    "    # Parse as ymd when SPECIMEN_COLLECTION_DATE is not a 5 digit numeric string - in a mdy (template submission) or ymd (Fulgent) format\n",
    "    !str_detect(SPECIMEN_COLLECTION_DATE, \"^[[:digit:]]{5}$\") ~ as.character(ymd(parse_date_time(SPECIMEN_COLLECTION_DATE, c(\"mdy\", \"ymd\"))))\n",
    "  )) %>%\n",
    "  \n",
    "# Transform and standardize GISAID_ID by submitter\n",
    "  ## Note: Aegis, Helix, Labcorp, and Quest are currently submitting via ELR (per 03.11.2022); Because ad-hoc corrections or other records from these labs are sometimes manually dropped into Submissions, we have retained the code to process these records\n",
    "  mutate(GISAID_ID = case_when(\n",
    "    \n",
    "    # Aegis submits a partial GISAID_ID with excess string at the end: remove '-BXXXXXX', append 'hCoV-19/' prefix, append '/[year of collection date]' suffix\n",
    "    SUBMITTING_LAB == \"Aegis\" & \n",
    "      replace_na(str_detect(toupper(SEQUENCE_STATUS), \"COMPLETE\"), FALSE) & \n",
    "      !is.na(GISAID_ID) \n",
    "    ~ paste0(\"USA/WA-CDC-\", str_extract(GISAID_ID, \"[^-]+\"), \"/\", year(SPECIMEN_COLLECTION_DATE)),\n",
    "\n",
    "    # Helix is submitting a full GISAID_ID so no edits necessary - however keeping this here in case future adjustments are necessary     \n",
    "    SUBMITTING_LAB == \"Helix\" & \n",
    "      replace_na(str_detect(toupper(SEQUENCE_STATUS), \"COMPLETE\"), FALSE) & \n",
    "      !is.na(GISAID_ID) \n",
    "    ~ paste0(GISAID_ID),\n",
    "    \n",
    "    # Labcorp submits a partial GISAID_ID: append 'hCoV-19/' prefix, append '/[year of collection date]' suffix\n",
    "    SUBMITTING_LAB == \"Labcorp\" & \n",
    "      replace_na(str_detect(toupper(SEQUENCE_STATUS), \"COMPLETE\"), FALSE) & \n",
    "      !is.na(GISAID_ID) \n",
    "    ~ paste0(\"USA/WA-CDC-\", GISAID_ID, \"/\", year(SPECIMEN_COLLECTION_DATE)),\n",
    "    \n",
    "    TRUE ~ GISAID_ID\n",
    "  ))"
   ],
   "id": "fcffb166-543f-4480-8318-9981a90af05a"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform matching\n",
    "\n",
    "CASE_IDs from all submitted records will be matched to the WDRS_ENTIRE table using SEQUENCE_CLINICAL_ACCESSION (exact match) and COLLECTION_DATE (+ or - 14 days). The inexact matching is conducted during the quality filters in a later chunk."
   ],
   "id": "a2f55159-c86b-46dc-98eb-d21a0dae9877"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Join WDRS CASE_ID to submissions by exact matching on sequence accession number\n",
    "records_matched <- left_join(records_mut, \n",
    "                           wdrs_ent,\n",
    "                           by = c(\"LAB_ACCESSION_ID\" = \"FILLER__ORDER__NUM__WDRS\"),\n",
    "                            na_matches = \"never\")\n",
    "\n",
    "# Extract unique records\n",
    "  ## Necessary step because CASE_ID can be repeated multiple times in WDRS. The join process can generate duplicate rows where each variable value is repeated.\n",
    "  ## This step identifies duplicates using all variables aside from column_label, which contains the filepath of each record and could falsely indicate that a record is unique if it was sent in two separate files\n",
    "records_matched_dedup <- records_matched[!duplicated(records_matched[,2:14]),]"
   ],
   "id": "5b477fa1-7315-47af-8da1-a49bd8b6dcc5"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The records_matched_dedup dataframe retains unmatched records in addition to those that were matched. Unmatched records are identifiable by the fact that they have not been assigned a CASE_ID. Next steps will divert the unmatched records to fuzzy matching (if they contain the requisite demographic information) or keep_na. The matched records will then be compiled into the initial roster and undergo subsequent QA checks.\n",
    "\n",
    "# Extract records for fuzzy matching\n",
    "\n",
    "The next code chunk scans the unnmatched observations and extracts those eligible for fuzzy matching based on the following criteria:\n",
    "\n",
    "Ineligible for fuzzy matching, proceed towards roster & keep_na: \\* CASE_ID is not missing (i.e. successfully matched) *OR* \\* All demographic columns are NA: FIRST_NAME, LAST_NAME, MIDDLE_NAME, DOB\n",
    "\n",
    "To fuzzy matching: \\* CASE_ID is missing (i.e. unmatched) *AND* \\* All demographic columns are not NA: FIRST_NAME, LAST_NAME, MIDDLE_NAME, DOB"
   ],
   "id": "b7a1ec80-504d-4625-b79a-67871b497b43"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compile matched records AND unmatched records without demographic information\n",
    "roster_initial <- records_matched_dedup[\n",
    "  (!is.na(records_matched_dedup$CASE_ID)) |\n",
    "  (rowSums(is.na(records_matched_dedup[, c(9:12)])) == 4)\n",
    ", ]\n",
    "\n",
    "# Extract unmatched records with demographic information for fuzzy matching\n",
    "fuzzy_match <- records_matched_dedup[\n",
    "  (is.na(records_matched_dedup$CASE_ID)) &\n",
    "  (!rowSums(is.na(records_matched_dedup[, c(9:12)])) == 4)\n",
    ", ]\n",
    "\n",
    "# If records have been flagged for fuzzy matching, transform and generate output\n",
    "if(nrow(fuzzy_match) > 0) {\n",
    "fuzzy_match_transform <- fuzzy_match %>%\n",
    "  select(-c(14:15)) \n",
    "}"
   ],
   "id": "f72d7b3b-4750-4b9d-9c99-c754116feb0c"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare roster\n",
    "\n",
    "## Format roster\n",
    "\n",
    "The next chunk transforms the roster_initial records into the output format used for rostering sequencing results. This dataframe contains matched records AND unmatched records without demographic information. Once mutated into the final roster format, the latter can be diverted to the keep_na list."
   ],
   "id": "7371f1e1-8e03-4850-abb6-bff5bb1db21e"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if(nrow(roster_initial) > 0) {\n",
    "  roster_initial_transform <- roster_initial %>%\n",
    "    \n",
    "  # mutate CASE_ID, as number\n",
    "  mutate(CASE_ID = as.numeric(CASE_ID)) %>%\n",
    "    \n",
    "  # mutate SEQUENCE_SGTF, populate as blank\n",
    "  mutate(SEQUENCE_SGTF = \"\") %>%\n",
    "    \n",
    "  # mutate SEQUENCE_SPECIMEN, populate as \"YES\"\n",
    "  mutate(SEQUENCE_SPECIMEN = \"YES\") %>%\n",
    "    \n",
    "  # mutate SEQUENCE_REASON, uppercase\n",
    "  mutate(SEQUENCE_REASON = toupper(SEQUENCE_REASON)) %>%\n",
    "    \n",
    "  # mutate SEQUENCE_DATE, populate as blank\n",
    "  mutate(SEQUENCE_DATE = \"\") %>%\n",
    "    \n",
    "  # mutate SUBMITTING_LAB, populate with corresponding values\n",
    "  mutate(SEQUENCE_LAB = case_when(\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"AEGIS\") ~ \"Aegis\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"ALTIUS\") ~ \"Altius\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"ASU\") ~ \"ASU\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"ATLAS\") ~ \"Atlas Genomics\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"BOISE VA\") ~ \"Boise VA\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"FULGENT\") ~ \"Fulgent Genetics\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"GRAVITY\") ~ \"Gravity\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"HELIX\") ~ \"Helix\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"INFINITY\") ~ \"Infinity\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"KAISER\") ~ \"KP WA Research Inst\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"LABCORP\") ~ \"Labcorp\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"LAURING LAB\") ~ \"Lauring Lab\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"NW GENOMICS\") ~ \"NW Genomics\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"QUEST\") ~ \"Quest\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"USAFSA\") ~ \"USAFSA\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"UNITED STATES AIR FORCE SCHOOL OF AEROSPACE\") ~ \"USAFSAM\",\n",
    "    str_detect(toupper(SUBMITTING_LAB), \"UW VIROLOGY\") ~ \"UW Virology\",\n",
    "    TRUE ~ SUBMITTING_LAB)) %>%\n",
    "    \n",
    "  # mutate SEQUENCE_STATUS, uppercase\n",
    "  # mutate(SEQUENCE_STATUS = toupper(SEQUENCE_STATUS)) %>%\n",
    "  # mutate(SEQUENCE_STATUS = if_else(PANGO_LINEAGE == \"Unassigned\", \"LOW QUALITY\", toupper(SEQUENCE_STATUS))) %>%\n",
    "    mutate(SEQUENCE_STATUS = case_when(\n",
    "      PANGO_LINEAGE == \"Unassigned\" ~ \"LOW QUALITY\", \n",
    "      TRUE ~ toupper(SEQUENCE_STATUS))) %>%\n",
    "    \n",
    "  # mutate SEQUENCE_REPOSITORY, populate with \"GISAID\"\n",
    "  mutate(SEQUENCE_REPOSITORY = \"GISAID\") %>% \n",
    "    \n",
    "  # mutate SEQUENCE_ACCESSION, remove \"hCoV-19/\" from GISAID_ID\n",
    "  mutate(SEQUENCE_ACCESSION = str_replace(GISAID_ID, \"hCoV-19/\", \"\")) %>%\n",
    "  \n",
    "  # mutate SEQUENCE_VARIANT_OPEN_TEXT, populate with PANGO_LINEAGE\n",
    "  mutate(SEQUENCE_VARIANT_OPEN_TEXT = case_when(\n",
    "    toupper(PANGO_LINEAGE) %in% c(lineages_list, \"INVALID\") ~ PANGO_LINEAGE,\n",
    "    TRUE ~ \"\"\n",
    "  )) %>% \n",
    "    \n",
    "  # mutate SEQUENCE_CLINICAL_ACCESSION, populate with LAB_ACCESSION_ID\n",
    "  mutate(SEQUENCE_CLINICAL_ACCESSION = LAB_ACCESSION_ID) %>%\n",
    "    \n",
    "  # mutate SEQUENCE_SPECIMEN_COLLECTION_DATE, populate in mdy format\n",
    "  mutate(SEQUENCE_SPECIMEN_COLLECTION_DATE = as.character(format(as.Date(roster_initial$SPECIMEN_COLLECTION_DATE), \"%m/%d/%Y\"))) %>%\n",
    "    \n",
    "  # mutate SEQUENCE_NOTES\n",
    "  mutate(SEQUENCE_NOTES = case_when(\n",
    "    # when PANGO_LINEAGE is not NA, populate into a message with PANGO_LINEAGE and date of processing\n",
    "    !is.na(PANGO_LINEAGE) ~ paste0(\"Lineage identified as \", PANGO_LINEAGE, \" on \", today(), \". Lineage assignments may change over time.\"),\n",
    "    # when PANGO_LINEAGE is not NA, populate as blank\n",
    "    is.na(PANGO_LINEAGE) ~ \"\"\n",
    "    )) %>% \n",
    "    \n",
    "  # mutate SEQUENCE NOTES for Unassigned records\n",
    "    mutate(SEQUENCE_NOTES = if_else(!is.na(SEQUENCE_VARIANT_OPEN_TEXT) & SEQUENCE_VARIANT_OPEN_TEXT!= \"Unassigned\", SEQUENCE_NOTES, NULL)) %>%\n",
    "\n",
    "  # mutate SEQUENCE_REVIEWED, populate as blank\n",
    "  mutate(SEQUENCE_REVIEWED = \"\") %>%\n",
    "    \n",
    "  # mutate Case.Note, populate with message\n",
    "  mutate(Case.Note = \"External data question package updated by COVID19 Sequencing Roster.\") %>% \n",
    "    \n",
    "  # select the variables required in the roster\n",
    "  select(CASE_ID, \n",
    "         SEQUENCE_SGTF, \n",
    "         SEQUENCE_SPECIMEN, \n",
    "         SEQUENCE_REASON, \n",
    "         SEQUENCE_DATE,\n",
    "         SEQUENCE_LAB, \n",
    "         SEQUENCE_STATUS,\n",
    "         SEQUENCE_REPOSITORY, \n",
    "         SEQUENCE_ACCESSION, \n",
    "         SEQUENCE_VARIANT_OPEN_TEXT,\n",
    "         SEQUENCE_CLINICAL_ACCESSION, \n",
    "         SEQUENCE_SPECIMEN_COLLECTION_DATE,\n",
    "         SEQUENCE_NOTES, \n",
    "         SEQUENCE_REVIEWED, \n",
    "         Case.Note\n",
    "  )\n",
    "}"
   ],
   "id": "8869f1d1-09c7-456d-8314-1ad23acb434a"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract records for keep_na\n",
    "\n",
    "The next code chunk examines the preliminary roster and removes records for keep_na. Records that qualify for keep_na are unmatched or do not contain sufficient data to be included in the final sequencing roster – they will be retained in the keep_na running list and re-processed by the keep_na script at a future date.\n",
    "\n",
    "Records designated for keep_na are identified using the following criteria:\n",
    "\n",
    "-   CASE_ID is missing\n",
    "\n",
    "-   SEQUENCE_CLINICAL_ACCESSION is missing\n",
    "\n",
    "-   SEQUENCE_STATUS is COMPLETE and SEQUENCE_ACCESSION is not missing\n",
    "\n",
    "-   SEQUENCE_STATUS is FAILED, NOT DONE, LOW QUALITY, or HIGH CT *and* SEQUENCE_ACCESSION is missing\n",
    "\n",
    "    *AND*\n",
    "\n",
    "-   SEQUENCE_STATUS is not PENDING\n",
    "\n",
    "The keep_na records are then checked against the WDRS_FLATTENED table and the running keep_na file to exclude rostered or previously processed records. A variable indicating the date is added to determine the retention time for these records once added to the keep_na file.\n",
    "\n",
    "The records that are not diverted to keep_na will be sent to undergo final pre-roster QA checks in subsequent chunks."
   ],
   "id": "63e1675e-3d96-4efc-880a-5ce9261c07c7"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Applies a series of logic checks to filter down records don't meet criteria to the keep_na file (often missing CASE_ID)\n",
    "keep_na <- roster_initial_transform[\n",
    "  \n",
    "    # CASE_ID is NA or\n",
    "  ((is.na(roster_initial_transform$CASE_ID)) |\n",
    "     \n",
    "    # SEQUENCE_CLINICAL_ACCESSION is NA or \n",
    "    (is.na(roster_initial_transform$SEQUENCE_CLINICAL_ACCESSION)) |\n",
    "     \n",
    "    # must NOT meet ANY of the conditions to be rostered: \n",
    "      ## (SEQUENCE_STATUS is COMPLETE & SEQUENCE_ACCESSION is not NA) or\n",
    "      ## (SEQUENCE_STATUS is FAILED & SEQUENCE_ACCESSION is NA) or\n",
    "      ## (SEQUENCE_STATUS is NOT DONE & SEQUENCE_ACCESSION is NA) or\n",
    "      ## (SEQUENCE_STATUS is LOW QUALITY & SEQUENCE_ACCESSION is NA) or\n",
    "      ## (SEQUENCE_STATUS is HIGH CT & SEQUENCE_ACCESSION is NA)\n",
    "    !((roster_initial_transform$SEQUENCE_STATUS == \"COMPLETE\") & (!is.na(roster_initial_transform$SEQUENCE_ACCESSION)) |\n",
    "    (roster_initial_transform$SEQUENCE_STATUS == \"FAILED\") & (is.na(roster_initial_transform$SEQUENCE_ACCESSION)) |\n",
    "    (roster_initial_transform$SEQUENCE_STATUS == \"NOT DONE\") & (is.na(roster_initial_transform$SEQUENCE_ACCESSION)) |\n",
    "      \n",
    "      # (roster_initial_transform$SEQUENCE_STATUS == \"LOW QUALITY\") & (is.na(roster_initial_transform$SEQUENCE_ACCESSION)) | \n",
    "    # Check if there are any records with LOW QUALITY that sill have a SEQUENCE_ACCESSION\n",
    "    ((roster_initial_transform$SEQUENCE_STATUS == \"LOW QUALITY\") & (roster_initial_transform$SEQUENCE_VARIANT_OPEN_TEXT == \"Unassigned\"))  |\n",
    "      \n",
    "      \n",
    "    (roster_initial_transform$SEQUENCE_STATUS == \"HIGH CT\") & (is.na(roster_initial_transform$SEQUENCE_ACCESSION))\n",
    "    )\n",
    "  ) &\n",
    "    \n",
    "  # SEQUENCE_STATUS is not PENDING \n",
    "  (roster_initial_transform$SEQUENCE_STATUS != \"PENDING\")  \n",
    "  , ]\n",
    "\n",
    "# Filter out records that exist in the WDRS_FLATTENED table or in the running keep_na file, then add a variable containing date processed\n",
    "keep_na_final <- keep_na[\n",
    "  # SEQUENCE_ACCESSION does not already exist within the WDRS_FLATTENED table and\n",
    "  (!(keep_na$SEQUENCE_CLINICAL_ACCESSION %in% wdrs_sca_flat_values)) &\n",
    "    \n",
    "  # SEQUENCE_CLINICAL_ACCESSION does not already exist within the keep_na  \n",
    "  (!(keep_na$SEQUENCE_CLINICAL_ACCESSION %in% keep_na_running$SEQUENCE_CLINICAL_ACCESSION)) \n",
    "  , ] %>%\n",
    "  \n",
    "  # Add a variable documenting the date that the record was processed\n",
    "  mutate(DATE_PROCESSED = as.character(today()))"
   ],
   "id": "a0463a79-4aa6-4cca-b75e-03dc4982036f"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform quality checks\n",
    "\n",
    "This chunk applies a series of logic checks to identify records that meet criteria to be uploaded and flag those that need manual review (For_Review/to_process). These QA checks are performed by the custom functions found in quality_filters.R."
   ],
   "id": "ef2db1e6-1aa1-4461-9be1-1f1b02848f43"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "roster_initial_transform_qa <- roster_initial_transform %>%\n",
    " # Add COLLECTION_DATE_WDRS column needed for quality filters\n",
    "  mutate(COLLECTION_DATE_WDRS = as.character(format(as.Date(roster_initial$SPECIMEN__COLLECTION__DTTM__WDRS), \"%m/%d/%Y\"))) %>%\n",
    " # Remove keep_na records\n",
    "  anti_join(keep_na)  \n",
    "  \n",
    "# Applies a series of logic checks to filter down records that meet the criteria to be uploaded into the WDRS_FLATTENED table\n",
    "temp_sub_quality <- roster_filters(\n",
    "                      roster_initial_transform_qa, \n",
    "                      lab_vars, \n",
    "                      wdrs_flat$CDC_N_COV_2019_SEQUENCE_ACCESSION_NUMBER, \n",
    "                      wdrs_flat$CDC_N_COV_2019_SEQUENCE_CLINICAL_ACCESSION_NUMBER, \n",
    "                      lineages_list)"
   ],
   "id": "a0043cd5-f994-4e58-aa01-8a4892c9b838"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create final roster"
   ],
   "id": "9e7ef904-5cf4-4626-ac49-4bbc3facc09c"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove any rows that failed quality checks to create final roster\n",
    "roster_final <- select(temp_sub_quality, -COLLECTION_DATE_WDRS) %>%\n",
    "  filter(sum==0) %>% \n",
    "  select(CASE_ID:Case.Note)"
   ],
   "id": "81d44e8b-9069-4452-9bb9-7e46a09e8133"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract records for manual review"
   ],
   "id": "f87ad533-990d-47ab-afbc-84b9fe80dbf6"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract records that triggered quality filter errors to be sent to the For_Review folder\n",
    "for_review <- filter(temp_sub_quality, sum>0) %>%\n",
    "  filter(SEQUENCE_STATUS != \"FAILED\" &\n",
    "         SEQUENCE_STATUS != \"LOW QUALITY\") %>%\n",
    "  filter(!(SEQUENCE_ACCESSION %in% keep_na$SEQUENCE_ACCESSION))"
   ],
   "id": "2675dcc7-b7e9-411c-b408-9b77897d447f"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------"
   ],
   "id": "56545317-359d-4f69-8b3a-b9a3268c3147"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Built-in stop before moving any files\n",
    "stop(\"Stop script before any files are moved. Opportunity to review file validity, final roster, and records sent to keep_na or fuzzy_match.\")"
   ],
   "id": "8a700647-ebf8-4d23-a3b4-02084dbb01a5"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate output\n",
    "\n",
    "For all the output files above, create vectors containing the filepaths and write all output files to their respective destination"
   ],
   "id": "0c079554-ef67-4525-b3e9-e9941fbcffee"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize vectors to contain the filepaths for all output files in the chunks above\n",
    "roster_final_filepath <- c()\n",
    "fuzzy_match_filepath <- c()\n",
    "for_review_filepath <- c()\n",
    "\n",
    "# If there are records in roster_final, output file as a '.csv'\n",
    "if(nrow(roster_final > 0)) {\n",
    "  # create filepath\n",
    "  roster_final_filepath <- paste0(\"write_roster_here/\", \"Template_Submitters_\", format(now(), \"%Y-%m-%d-%H%M%S\"), \".csv\")\n",
    "  # write to csv\n",
    "  write_csv(roster_final, roster_final_filepath , na = \"\")\n",
    "}\n",
    "\n",
    "# If there are records in fuzzy_match, output file as a '.csv'\n",
    "if (nrow(fuzzy_match) > 0) {\n",
    "  # create filepath\n",
    "  fuzzy_match_filepath <- paste0(\"Submissions/Fuzzy_Match/\", \"Template_Submitters_Fuzzy_Match_Required_\", format(now(), \"%Y-%m-%d-%H%M%S\"), \".csv\")\n",
    "  # write to csv\n",
    "  write_csv(fuzzy_match_transform, fuzzy_match_filepath, na = \"\")\n",
    "}\n",
    "\n",
    "# If there are records in for_review, output file as a '.csv'\n",
    "if(nrow(for_review > 0)) {\n",
    "  # create filepath\n",
    "  for_review_filepath <- paste0(\"For_Review/to_process/\", \"For_Review_Template_Submitters_\", format(now(), \"%Y-%m-%d-%H%M%S\"), \".csv\")\n",
    "  # write to csv\n",
    "  write_csv(for_review, for_review_filepath, na = \"\")\n",
    "}\n",
    "\n",
    "# If there are records in keep_na_final, append the records to the running keep_na file \n",
    "if(nrow(keep_na_final) > 0) {\n",
    "  keep_na_final %>%\n",
    "    write_csv(paste0(\"keep_na/keep_na.csv\"), na = \"\", append = TRUE)\n",
    "} "
   ],
   "id": "17256245-3fb4-485d-ba9d-30c7bdb025e1"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ensure that records were appended to the keep_na file by comparing the number of records in the past and current versions of the object. The difference should equal the nrow of the keep_na_final dataframe created during this script run. If the number of records do not match, output the keep_na_final object to a holding folder to be appended later."
   ],
   "id": "083e52d7-1308-4eb7-af1e-857690b1bb11"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if(nrow(keep_na_final) > 0) {\n",
    "\n",
    "  # Read in the updated keep_na \n",
    "  keep_na_running_update <- read_csv(\"keep_na/keep_na.csv\",\n",
    "                   col_types = cols(.default = \"c\"),\n",
    "                   na = c(\"\", \"NA\", \"N/A\")) \n",
    "  \n",
    "  # If the difference in nrow between the previous iteration of keep_na and the updated version does not equal the number of rows in keep_na_final,\n",
    "  if ( (!( (nrow(keep_na_running_update)) - (nrow(keep_na_running)) ) == nrow(keep_na_final) ) ) {\n",
    "    # Output keep_na_final as a separate file into a holding folder to ensure these data are reviewed\n",
    "    keep_na_final %>%\n",
    "      write_csv(paste0(\"keep_na/Add_Holding/\", \"Template_Submitters_\", format(now(), \"%Y-%m-%d-%H%M%S\"), \".csv\"), na = \"\")\n",
    "  }\n",
    "}"
   ],
   "id": "75e33a5e-b2a1-4459-ad23-1282511c4759"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Email receipt\n",
    "\n",
    "Compile an email message by concatenating the vectors created in previous chunks containing relevant messages."
   ],
   "id": "12fa61cc-62e7-49e7-bfef-c2a0c689f233"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare email message\n",
    "\n",
    "## Valid/invalid files\n",
    "valid_files_message <- c()\n",
    "valid_files_message <-  paste0(valid_files_subset[,1], collapse=\"\\n\")\n",
    "final_valid_files_message <- paste0(\"The following file(s) are valid and have been processed:\\n\", valid_files_message, collapse = \"\\n\\n\")\n",
    "\n",
    "## Empty files\n",
    "empty_submissions_message <-  paste0(empty_submissions, collapse=\"\\n\")\n",
    "\n",
    "# Final message\n",
    "  ## If the script makes it to this point it will have valid files that were processed -- start with this message and append other potential details\n",
    "message <- paste0(final_valid_files_message, \"\\n\\n\", \"A total of \", nrow(roster_final), \" records were sent to write_roster_here.\")\n",
    "\n",
    "  ## If there were invalid files, append it to the message\n",
    "if ((nrow(invalid_files_subset) > 0)) {\n",
    "  message <- paste0(message,\"\\n\\n\" , final_invalid_files_message)\n",
    "}\n",
    "\n",
    "  ## If there were empty files, append it to the message\n",
    "if ((length(empty_submissions) > 0)) {\n",
    "  message <- paste0(message, \"\\n\\n\", \"The following file(s) were received but contained no data:\", \"\\n\", empty_submissions_message)\n",
    "}\n",
    "\n",
    "  ## If there were records that required fuzzy matching, append it to the message\n",
    "if ((nrow(fuzzy_match) > 0)) {\n",
    "  message <- paste0(message, \"\\n\\n\", \"There are a subset of records from the valid files that were processed that require fuzzy matching. These records do not match to an event in WDRS by an accession ID but do contain viable demographic information:\", \"\\n\", fuzzy_match_filepath)\n",
    "}\n",
    "\n",
    "message <- paste0(message, \"\\n\\n\", \"Note: This is an automated message. Please enable your outlook to include extra line breaks to view this message in its proper format.\")"
   ],
   "id": "f1b0b548-dd41-40a6-9873-a2e257b60888"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Email details\n",
    "email_from <- \"\"\n",
    "email_to <- \"\"\n",
    "email_subj <- \"SEQUENCING - Template Submitters Automated Email\"\n",
    "email_body <- paste0(\"Template submission(s) have been processed \", today(), \".\", \"\\n\\n\", message)\n",
    "\n",
    "# Send email\n",
    "sendmailR::sendmail(from = email_from,\n",
    "                    to = email_to,\n",
    "                    subject = email_subj,\n",
    "                    msg = email_body,\n",
    "                    headers= list(\"Reply-To\" = email_from),\n",
    "                    control = list(smtpServer = \"\")\n",
    "                    )"
   ],
   "id": "9ab58e74-938a-46df-a361-fbc7a4154812"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Move processed files"
   ],
   "id": "b9584be0-d1da-4b1e-8f45-714a57387512"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create vector of valid files to move\n",
    "move_valid_files <- valid_files_subset[[1]]\n",
    "# Create vector of valid files and empty files to move\n",
    "move_files <- c(move_valid_files, empty_submissions)\n",
    "\n",
    "# For every file processed, copy from Submissions to the corresponding Completed_Submissions folder and delete original\n",
    "  ## All files should be created successfully, but the script may be unable to delete some files that it considers in use by the R environment. These will be removed by Chunk 10 in the next run prior to identifying new submissions for processing\n",
    "for (i in 1:length(move_files)) {\n",
    "  \n",
    "  ## Adhoc\n",
    "  if (str_detect(move_files[i], \"Adhoc\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/Adhoc\")\n",
    "    unlist(move_files[i])\n",
    "\n",
    "  ## Aegis\n",
    "  } else if (str_detect(move_files[i], \"Aegis\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/Aegis\")\n",
    "    unlist(move_files[i])\n",
    "  \n",
    "  ## Altius\n",
    "  } else if (str_detect(move_files[i], \"Altius\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/Altius\")\n",
    "    unlist(move_files[i])\n",
    "    \n",
    "  ## ASU\n",
    "  } else if (str_detect(move_files[i], \"ASU\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/ASU\")\n",
    "    unlist(move_files[i])\n",
    "\n",
    "  ## Fulgent Genetics\n",
    "  } else if (str_detect(move_files[i], \"Fulgent_Genetics\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/Fulgent_Genetics\")\n",
    "    unlist(move_files[i])\n",
    "    \n",
    "  ## Gravity Diagnostics\n",
    "  } else if (str_detect(move_files[i], \"Gravity Diagnostics\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/Gravity_Diagnostics\")\n",
    "    unlist(move_files[i])\n",
    "    \n",
    "  ## Helix  \n",
    "  } else if (str_detect(move_files[i], \"Helix\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/Helix\")\n",
    "    unlist(move_files[i])\n",
    "    \n",
    "  ## Infinity Biologix\n",
    "  } else if (str_detect(move_files[i], \"Infinity_Biologix\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/Infinity_Biologix\")\n",
    "    unlist(move_files[i])\n",
    "\n",
    "  ## Kaiser\n",
    "  } else if (str_detect(move_files[i], \"Kaiser\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/Kaiser\")\n",
    "    unlist(move_files[i])\n",
    "  \n",
    "  ## Labcorp\n",
    "  } else if (str_detect(move_files[i], \"Labcorp\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/Labcorp\")\n",
    "    unlist(move_files[i])\n",
    "\n",
    "  ## Lauring Lab\n",
    "  } else if (str_detect(move_files[i], \"Lauring_Lab\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/Lauring_Lab\")\n",
    "    unlist(move_files[i])\n",
    "  \n",
    "  ## NW Genomics\n",
    "  } else if (str_detect(move_files[i], \"NW_Genomics\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/NW_Genomics\")\n",
    "    unlist(move_files[i])\n",
    "    \n",
    "  ## United States Air Force\n",
    "  } else if (str_detect(move_files[i], \"UnitedStatesAirForceSchoolofAerospace\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/USAFSAM\")\n",
    "    unlist(move_files[i])\n",
    "    \n",
    "  ## UW Virology\n",
    "  } else if (str_detect(move_files[i], \"UW_Virology\")) {\n",
    "    file.copy(move_files[i], \"Completed_Submissions/UW_Virology\")\n",
    "    unlist(move_files[i])\n",
    "  }\n",
    "}\n",
    "\n",
    "# Update completed files list for next time\n",
    "write_csv(as.data.frame(move_files), \"Completed_Submissions/Template_Submitters/Template_Submitters_File_Completed.csv\")"
   ],
   "id": "9dddbdda-9304-4ee4-8cb5-05e093c7b40e"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
